File: src/main/java/shenefelt/libraryManager/Library.java
package shenefelt.libraryManager;

import java.util.*;

import static java.lang.System.out;

public class Library
{

    /* ---- BEGIN LIBRARY ----*/
    private HashMap<Integer, BookShelf> shelfMap = null;

    public Library() {}


    /**
     * Add a shelf to the library
     * @param bookShelf the BookShelf that will be added to the library
     */
    public void addBookShelf(BookShelf bookShelf)
    {
        if (shelfMap == null)
            shelfMap = new HashMap<>();

        shelfMap.put(bookShelf.getID(), bookShelf);
    }

    /**
     * Get a BookShelf of a given ID number
     * @param shelfId the shelf ID of the shelf.
     * @return the BookShelf object that has a matching shelf ID
     */
    public BookShelf getBookShelf(int shelfId)
    {
        return shelfMap.get(shelfId);
    }

    /**
     * Print a list of all Shelves in the Library - Includes Shelf IDs
     */
    public void printShelfMap()
    {
        for(Map.Entry<Integer, BookShelf> shelf : shelfMap.entrySet())
            out.println("Shelf ID: " + shelf.getKey() + ", Book Shelf: " + shelf.getValue());
    }

    /**
     * Display the total number of books & shelves in the library
     */
    public void displayTotals()
    {
        if(shelfMap == null) {return;}

        out.println("Total Books: " + getTotalBooks());
        out.println("Total Shelves: " + shelfMap.size());
        out.println();
    }

    /**
     * List each book in the library
     *  List BookID as well.
     */
    public void listAllBooks()
    {
        if(shelfMap.isEmpty())
        {
            out.println("No books found in the library :(");
        }


    }

    /**
     * Calls the recursive counting method
     * @return the sum of all books in the library
     */
    public int getTotalBooks()
    {
        return countBooksR(new ArrayList<>(shelfMap.values()), 0, 0);
    }

    /*
        Private methods
     */

    /**
     * Recurse the shelf map to get a total count of all books
     * @param shelves the BookShelf objects we wish to count
     * @param index the current position in the list
     * @param sum the total number of all books
     * @return the sum of all books on the shelves, respecting the terminating steps
     */
    private int countBooksR(List<BookShelf> shelves, int index, int sum)
    {
        if(shelves == null || index < 0)
            return 0;

        if (index >= shelves.size())
            return sum;

        return countBooksR(shelves, index + 1, (sum + shelves.get(index).getNumberOfBooks()));
    }


    /**
     * Search the library for a book by it's given title
     * @param resultList an ArrayList of all matching titles.
     * @param identity the title of the book being searched for.
     * @param shelfCollection the map of all shelves to search
     * @return true if the title is found else false.
     */
    private boolean searchLibraryForTitle (ArrayList<Book> resultList, String identity,
                                           Hashtable<Integer, BookShelf> shelfCollection)
    {
        boolean found = false;
        Book result = null;
        int shelfID = 0;

        for(BookShelf s : shelfCollection.values())
        {
            result = s.findBookByTitle(identity);
            shelfID = s.getID(); // I want the search results to display what shelf a book was found on
            if (result != null)
            {
                resultList.add(result);
                found = true;
            }
        }

        return found;
    }


}

File: src/main/java/shenefelt/libraryManager/kotlin/Library.kt
package shenefelt.libraryManager.kotlin

import shenefelt.libraryManager.BookShelf
import java.util.*

class Library
{


    private var shelfMap: MutableMap<Int, BookShelf>? = null
    private val books = mutableSetOf(Book())


    val hasBooks: Boolean
        get() = !shelfMap.isNullOrEmpty()

    fun addBookShelf(bookShelf: BookShelf) {
        if (shelfMap == null)
            shelfMap = HashMap()
        shelfMap?.put(bookShelf.getID(), bookShelf)


    }

    fun addBook(newBook: Book): Boolean {
        val book = Book()
        book.setTotalPages(1500)
        println(book.getGenre())

        val shelf = if(shelfMap.isNullOrEmpty())
            null
        else
            getOpenShelf()

        return if (shelf !== null) {
            shelf.addBook(newBook)
        } else {
            false
        }

    }

    private fun getOpenShelf(): BookShelf? {
        shelfMap?.values?.forEach { bookShelf ->
            if (bookShelf.hasCapacity())
                return bookShelf
        }
        return null
    }

    fun getBookShelf(shelfId: Int): BookShelf? = shelfMap?.get(shelfId)

    fun printShelfMap() = shelfMap?.entries?.forEach { shelf ->
        println("Shelf ID: $${shelf.key}, Book Shelf: $${shelf.value}")
    }

    fun displayTotals() {
        if (shelfMap == null)
            return

        println("Total Books: $${getTotalBooks()}")
        println("Total Shelves: $${shelfMap?.size}")
        println()
    }

    fun listAllBooks()
    {
        if (shelfMap!!.isEmpty()) {
            println("No books found in the library :(")
        }
        // loop through shelves and list books
    }

    fun getTotalBooks(): Int = countBooksR(ArrayList(shelfMap!!.values), 0, 0)

    private fun countBooksR(shelves: List<BookShelf>,
                            index: Int,
                            sum: Int): Int {
        return if (shelves.isEmpty() || index < 0)
            0
        else if (index >= shelves.size)
            sum
        else
            countBooksR(shelves, index + 1,
                sum + shelves[index].getNumberOfBooks())
    }

    private fun searchLibraryForTitle(resultList: ArrayList<Book>,
                                      identity: String,
                                      shelfCollection: Hashtable<Int, BookShelf>): Boolean {

        var found = false
        var result: Book?
        var shelfID: Int

        for (s in shelfCollection.values) {
            result = s.findBookByTitle(identity)
            shelfID = s.getID()
            if (result != null) {
                resultList.add(result)
                found = true
            }
        }

        return found
    }
}